qiskit_nature/second_q/operators/polynomial_tensor.py

```diff
@@ -0,0 +1,119 @@
+
+"""Polynomial Tensor class"""
+
+from __future__ import annotations
+from typing import Dict
+from numbers import Number
+import numpy as np
+from qiskit.quantum_info.operators.mixins import (
+    LinearMixin,
+    AdjointMixin,
+    TolerancesMixin,
+)
+
+
+class PolynomialTensor(LinearMixin, AdjointMixin, TolerancesMixin):
+    """Polynomial Tensor class"""
+
+    def __init__(self, data: Dict[str, np.ndarray]):
+        if not isinstance(data, Dict):
+            raise TypeError(f"data {data} must be a Dictionary")
+
+        self._data = data
+
+        shapes = ()
+        for key, value in self._data.items():
+            shapes += value.shape
+            if len(value.shape) != len(key):
+                raise ValueError(
+                    f"data key {key} of length {len(key)} does not match "
+                    f"data value matrix of dimensions {value.shape}"
+                )
+            if len(set(value.shape)) != 1:
+                raise ValueError(
+                    f"For key {key}: dimensions of value matrix are not identical {value.shape}"
+                )
+
+        if len(set(shapes)) != 1:
+            raise ValueError("Dimensions of value matrices in data dictionary are not identical.")
+
+    def mul(self, other: complex):
+        """Scalar multiplication of PolynomialTensor with complex"""
+
+        if not isinstance(other, Number):
+            raise TypeError(f"other {other} must be a number")
+
+        prod_dict = {}
+        for key, matrix in self._data.items():
+            prod_dict[key] = matrix * other
+        return PolynomialTensor(prod_dict)
+
+    def _multiply(self, other):
+        return self.mul(other)
+
+    def add(self, other: PolynomialTensor):
+        """Addition of PolynomialTensors"""
+
+        if not isinstance(other, PolynomialTensor):
+            raise TypeError("Incorrect argument type: other should be PolynomialTensor")
+
+        sum_dict = self._data.copy()
+        for other_key, other_value in other._data.items():
+            if other_key in sum_dict.keys():
+                if other_value.shape == np.shape(sum_dict[other_key]):
+                    sum_dict[other_key] = np.add(other_value, sum_dict[other_key])
+                else:
+                    print("not same shape", other_value.shape, np.shape(sum_dict[other_key]))
+                    raise ValueError(
+                        f"For key {other_key} "
+                        f"corresponding data value of shape {np.shape(sum_dict[other_key])} "
+                        f"does not match other value matrix of shape {other_value.shape}"
+                    )
+            else:
+                sum_dict[other_key] = other_value
+
+        return PolynomialTensor(sum_dict)
+
+    def _add(self, other, qargs=None):
+        return self.add(other)
+
+    def __eq__(self, other):
+        """Check equality of PolynomialTensors"""
+
+        if self._data.keys() == other._data.keys():
```

Comment:
```suggestion
        if set(self._data.keys()) == set(other._data.keys()):
```
I think otherwise, this is dependent on the order of the keys